### DISCUSSION TITLE
Reading pkl objects remotely

### DISCUSSION THREAD

**Discussion by fhalde** [apple/pkl#645] [General]
Hello,

Let's say our platform has defined some standard Pkl class `Foo` to represent some entity in our platform named `Foo`
```
class Foo {
 name: String
}

```

Is it possible to write a Pkl configuration that reads these Foo objects exposed over an API ?

```
myfoo: Foo = read("http://localhost:9092/foo/1")
```

We are mainly unsure what should be the return format of the endpoint
- JSON
- Pcf Rendered Pkl object

I'm open for suggestions.

```
read("http://localhost:9092/foo/1")
=> new Resource { uri = "file:stream.pkl"; text = "bird {\n hello = 1\n}\n"; base64 = "YmlyZCB7CiBoZWxsbyA9IDEKfQo=" }
```
Could this text be parsed into a Pkl class instance?

**Comment by fhalde** [comment]
Alternatively, would it make sense to `import "http://localhost:9092/foo/1"` ?

**Comment by StefMa** [comment]
You can put your base pkl file somewhere on the internet and then just import via https.
See also https://pkl-lang.org/main/current/language-reference/index.html#import-module

Is that what you're looking for? 🤔

**Reply by fhalde** [reply]
ya i'm thinking in the same lines

1. `master.pkl` => declares all the known types in the platform.
2. users configure instances of all the known types as `.pkl` files and make these pkl files addressable over an endpoint.
3. application writers, `import` these user generated pkl files using their well-known-uri to produce their final app configuration.

**Comment by bioball** [comment]
Both `import` and `read` are able to interact with HTTPS endpoints. So, both of these work, assuming that the remote resource exists:

```pkl
import "https://example.com/path/to/remote/module.pkl"

remoteResource: Resource = read("https://www.example.com/path/to/remote/resource.json")
```

Note that resources fetch over HTTP(S) do not get cached across evaluations. So, calling Pkl will result in an HTTP call every time.
Additionally, to ensure reproducible output within a single evaluation, Pkl will cache the result of any import/read by its URI per evaluator session.
